### PR TITLE
Search index checker

### PR DESCRIPTION
**Pull request by jcollins-g**
Add a search index checker for #1434 and make some enhancements to warnings for broken links.

The original issue #1434 would now result in more informative broken link errors:
```
 warning: dartdoc generated a broken link to: angular_components/ActivateItemOnKeyPressMixin/isRtl.html
    to element angular_components.MaterialDropdownSelectComponent.isRtl: (file:///usr/local/google/home/jcollins/dart/angular2_components/lib/src/model/a11y/keyboard_handler_mixin.dart:40:8)
    linked to from angular_components.MaterialDropdownSelectComponent.visible: (file:///usr/local/google/home/jcollins/dart/angular2_components/lib/src/components/material_select/material_dropdown_select.dart)
 warning: dartdoc generated a broken link to: angular_components/ActivateItemOnKeyPressMixin/handleEndKey.html
    to element angular_components.MaterialDropdownSelectComponent.handleEndKey: (file:///usr/local/google/home/jcollins/dart/angular2_components/lib/src/model/a11y/keyboard_handler_mixin.dart:120:8)
    linked to from angular_components.MaterialDropdownSelectComponent.onFocus: (file:///usr/local/google/home/jcollins/dart/angular2_components/lib/src/components/material_select/material_dropdown_select.dart:207:8)
```

It would appear that since 0.12.0 the search index being filled with incorrect hyperlinks is no longer happening, but here is output from simulated failures in the index.json link checker:

```
 warning: dartdoc generated a broken link to: angular_components/KeyUpBoundaryDirective/hashCode.html (from index.json), to element angular_components.KeyUpBoundaryDirective.hashCode: (file:///usr/local/google/home/jcollins/dart/all_sdks/1.24.0-dev.5.0/lib/core/object.dart)
 warning: dartdoc generated a broken link to: angular_components/MaterialInkTooltipComponent/tooltipRef.html (from index.json), to element angular_components.MaterialInkTooltipComponent.tooltipRef: (file:///usr/local/google/home/jcollins/dart/angular2_components/lib/src/components/material_tooltip/src/ink_tooltip.dart)
 warning: dartdoc generated a file not in the search index: angular_components/StringSelectionOptions-class.html (from index.json), from angular_components.StringSelectionOptions: (file:///usr/local/google/home/jcollins/dart/angular2_components/lib/src/model/selection/string_selection_options.dart:36:7)
 warning: dartdoc generated a file not in the search index: angular_components/TrackLayoutChangesMixin/noSuchMethod.html (from index.json), from angular_components.TrackLayoutChangesMixin.noSuchMethod: (file:///usr/local/google/home/jcollins/dart/all_sdks/1.24.0-dev.5.0/lib/core/object.dart:107:20)
```

### CODE REVIEW

**Comment by kevmoo** [comment]
Drive by: how is this tested?

**Comment by jcollins-g** [comment]
This is a test.

**Comment by jcollins-g** [comment]
More specifically: this is an enhancement to dartdoc's self-checking, and so it didn't make much sense to me to spend a great deal of time testing outside of manual checks that the check works.

Nevertheless, dartdoc in general needs improved testing and I have open issues on that.

File: lib/src/model.dart [hunk 1]
```diff
@@ -3092,6 +3097,8 @@ class Package implements Nameable, Documentable {
         break;
       case PackageWarning.brokenLink:
         warningMessage = 'dartdoc generated a broken link to: ${message}';
+        warnablePrefix = 'to element';
+        referredFromPrefix = 'linked to from';
```
Comment (devoncarew) [inline lib/src/model.dart]: `linked to from` => `referenced from`?

**Comment by devoncarew** [comment]
lgtm

Not as part of this PR, but you may want to make the location output a bit more terse. In ModelElement `String get elementLocation`, you could prefer a path to a uri if it was a file: uri, and you could use a relative path if the source was contained in the cwd. That would convert

```
file:///usr/local/google/home/jcollins/dart/angular2_components/lib/src/model/a11y/keyboard_handler_mixin.dart:40:8
``

to 

```
lib/src/model/a11y/keyboard_handler_mixin.dart:40:8
```

or perhaps

```
angular2_components/lib/src/model/a11y/keyboard_handler_mixin.dart:40:8
```

**Comment by kevmoo** [comment]
> This is a test.

Right, but it's a user-facing feature – which should have it's own tests. Just curious..

**Comment by jcollins-g** [comment]
@kevmoo I'd be happy to discuss my testing philosophy with you offline, feel free to drop something on my calendar.

**Comment by jcollins-g** [comment]
re: shortening URLs, that should be a lot more possible since they're all generated from the same place now, just haven't gotten to it yet.

**Pull request merged**